tutorralph/templates/ralph/apps/config/env

```diff
@@ -1,78 +1,18 @@
-# define them for convenience purpose during development, but they can be
-# passed as CLI options.
-
-# RALPH_BACKENDS__STORAGE__LDP__ENDPOINT=
-# RALPH_BACKENDS__STORAGE__LDP__APPLICATION_KEY=
-# RALPH_BACKENDS__STORAGE__LDP__APPLICATION_SECRET=
-# RALPH_BACKENDS__STORAGE__LDP__CONSUMER_KEY=
-# RALPH_BACKENDS__STORAGE__LDP__SERVICE_NAME=
-# RALPH_BACKENDS__STORAGE__LDP__STREAM_ID=
-
-# Swift storage backend
-
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_AUTH_URL=http://swift:35357/v3/
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_IDENTITY_API_VERSION=3
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_USER_DOMAIN_NAME=Default
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_PROJECT_DOMAIN_NAME=Default
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_TENANT_ID=cd238e84310a46e58af7f1d515887d88
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_TENANT_NAME=RegionOne
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_USERNAME=demo
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_PASSWORD=demo
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_REGION_NAME=RegionOne
-# RALPH_BACKENDS__STORAGE__SWIFT__OS_STORAGE_URL=http://swift:8080/v1/KEY_cd238e84310a46e58af7f1d515887d88/test_container
-
-# ES database backend
-
-RALPH_BACKENDS__DATABASE__ES__HOSTS=http://elasticsearch:9200
-RALPH_BACKENDS__DATABASE__ES__INDEX=statements
-RALPH_BACKENDS__DATABASE__ES__TEST_HOSTS=http://elasticsearch:9200
-RALPH_BACKENDS__DATABASE__ES__TEST_INDEX=test-index
-RALPH_BACKENDS__DATABASE__ES__TEST_FORWARDING_INDEX=test-index-2
-
-# MONGO database backend
-
-RALPH_BACKENDS__DATABASE__MONGO__COLLECTION=foo
-RALPH_BACKENDS__DATABASE__MONGO__DATABASE=statements
-RALPH_BACKENDS__DATABASE__MONGO__CONNECTION_URI=mongodb://mongo:27017/
-RALPH_BACKENDS__DATABASE__MONGO__TEST_COLLECTION=foo
-RALPH_BACKENDS__DATABASE__MONGO__TEST_FORWARDING_COLLECTION=foo-2
-RALPH_BACKENDS__DATABASE__MONGO__TEST_DATABASE=statements
-RALPH_BACKENDS__DATABASE__MONGO__TEST_CONNECTION_URI=mongodb://mongo:27017/
-
 # ClickHouse database backend
-RALPH_BACKENDS__DATABASE__CLICKHOUSE__HOST=clickhouse
-RALPH_BACKENDS__DATABASE__CLICKHOUSE__PORT=8123
-RALPH_BACKENDS__DATABASE__CLICKHOUSE__XAPI_DATABASE=xapi
-RALPH_BACKENDS__DATABASE__CLICKHOUSE__EVENT_TABLE_NAME=xapi_events_all
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__HOST={{CLICKHOUSE_HOST}}
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__PORT={{CLICKHOUSE_HTTP_PORT}}
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__XAPI_DATABASE={{OARS_XAPI_DATABASE}}
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__EVENT_TABLE_NAME={{OARS_RAW_XAPI_TABLE}}
 RALPH_BACKENDS__DATABASE__CLICKHOUSE__TEST_HOST=clickhouse
-RALPH_BACKENDS__DATABASE__CLICKHOUSE__USERNAME={{CLICKHOUSE_USERNAME}}
-RALPH_BACKENDS__DATABASE__CLICKHOUSE__PASSWORD={{CLICKHOUSE_PASSWORD}}
-
-# Sentry
-
-# RALPH_SENTRY_DSN=
-# RALPH_EXECUTION_ENVIRONMENT=development
-
-# Converter
-
-# RALPH_CONVERTER_EDX_XAPI_UUID_NAMESPACE=
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__USERNAME={{CLICKHOUSE_ADMIN_USER}}
+RALPH_BACKENDS__DATABASE__CLICKHOUSE__PASSWORD={{CLICKHOUSE_ADMIN_PASSWORD}}
```

Comment:
FYI @bmtcril -- I tried a bunch of different things to try to address this, but to get it working, I ended up just using the Clickhouse Admin user here :(

I tried:

* Using the `ralph auth` command-line tool to add the Ralph LRS user, because Ralph's startup text indicated that it was watching `/app` for changes, so I thought it would restart if the `/app/.ralph/auth.json` file changes. It didn't.
* Setting docker-compose `healthcheck` and `restart` statements for the ralph service based on successfully hitting the /xAPI/statements endpoint -- I couldn't get the ralph container to restart on failure.
* Creating a second container called `ralph_clickhouse` that `ralph` `depends_on`, which tried checking the clickhouse connection using the CLICKHOUSE_LMS user to indicate its success -- the best I could do was mark the `ralph` container as "unhealthy", but couldn't get anything to restart.

I still think the best actual option is to modify the docker deployment for Ralph to run it as a service that restarts itself until it's running properly, but since I couldn't figure out how to do it with the docker-compose container, that was out of scope here.

Reply:
Yeah I think the fact that Ralph wedges instead of stopping when it hits that error is problem 1, but I'm reconsidering having the services own their users instead of OARS to prevent these tangles going forward. Presumably even if they weren't being used in an OARS context they would need the same types of users. Still pondering on it, though.